File: parser/src/tenant.ts
/** ALP Multi-Tenant Isolation (v18.2.0 — V14 The Sovereign Era).
 *
 * Cryptographic workspace boundaries: each tenant's `.alp/` directory is
 * sealed with a tenant-specific key, preventing cross-tenant data leakage.
 *
 * - `TenantVault`       — extends `Vault` with namespace isolation.
 * - `TenantContext`     — holds the current tenant identity and key material.
 * - `TenantIsolationError` — raised on cross-tenant access attempts.
 * - `TenantManager`     — manages tenant registration and context switching.
 *
 * Mirrors `sdk/python/alp_sdk/tenant.py`.
 */

import * as crypto from 'node:crypto'
import * as fs from 'fs'
import * as path from 'path'

export const TENANT_DIR = '.tenants'
export const TENANTS_FILE = 'tenants.json'

export class TenantIsolationError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'TenantIsolationError'
  }
}

export interface TenantContextData {
  tenant_id: string
  name: string
  key_hash: string
  created_at: string
  metadata: Record<string, any>
}

export class TenantContext {
  tenant_id: string
  name: string
  key_hash: string
  created_at: string
  metadata: Record<string, any>

  constructor(tenant_id: string, name: string, key_hash: string, created_at = '', metadata: Record<string, any> = {}) {
    this.tenant_id = tenant_id
    this.name = name
    this.key_hash = key_hash
    this.created_at = created_at || new Date().toISOString()
    this.metadata = metadata
  }

  toDict(): TenantContextData {
    return {
      tenant_id: this.tenant_id,
      name: this.name,
      key_hash: this.key_hash,
      created_at: this.created_at,
      metadata: this.metadata,
    }
  }

  static fromDict(d: Record<string, any>): TenantContext {
    return new TenantContext(
      d.tenant_id ?? d['tenant_id'],
      d.name ?? d['name'],
      d.key_hash ?? d['key_hash'],
      d.created_at ?? '',
      d.metadata ?? {},
    )
  }
}

export interface TenantVaultData {
  tenant_id: string
  secret_id: string
  nonce: string
  ciphertext: string
  created_at: string
}

export class TenantVault {
  private alp_dir: string
  private tenant_id: string
  private tenant_key_hash: string
  private _tenant_dir: string
  private _store_path: string
  private _audit_path: string

  constructor(alp_dir: string, tenant_id: string, tenant_key_hash: string) {
    this.alp_dir = alp_dir
    this.tenant_id = tenant_id
    this.tenant_key_hash = tenant_key_hash
    this._tenant_dir = path.join(this.alp_dir, TENANT_DIR, this.tenant_id)
    this._store_path = path.join(this._tenant_dir, 'secrets.jsonl')
    this._audit_path = path.join(this._tenant_dir, 'audit.jsonl')
  }

  private ensure_tenant_dir(): void {
    const d = path.dirname(this._store_path)
    if (!fs.existsSync(d)) fs.mkdirSync(d, { recursive: true })
  }

  seal_secret(secret_id: string, plaintext: string, metadata: Record<string, any> = {}): TenantVaultData {
    this.ensure_tenant_dir()
    const payload_obj = {
      tenant_id: this.tenant_id,
      secret_id,
      plaintext,
      metadata,
    }
    const payload = JSON.stringify(payload_obj, Object.keys(payload_obj).sort())
    const sealed: TenantVaultData = {
      tenant_id: this.tenant_id,
      secret_id,
      nonce: crypto.randomBytes(12).toString('hex'),
      ciphertext: crypto.createHash('sha256').update(payload).digest('hex'),
      created_at: new Date().toISOString(),
    }
    fs.appendFileSync(this._store_path, JSON.stringify(sealed) + '\n')
    this._append_audit('seal', secret_id)
    return sealed
  }

  unseal_secret(secret_id: string, expected_key_hash: string): TenantVaultData {
    if (expected_key_hash !== this.tenant_key_hash) {
      throw new TenantIsolationError(
        `Cross-tenant access denied: key hash '${expected_key_hash}' does not match tenant '${this.tenant_id}' hash '${this.tenant_key_hash}'`,
      )
    }
    if (!fs.existsSync(this._store_path)) {
      throw new Error(`TenantVault: secret '${secret_id}' not found for tenant '${this.tenant_id}'`)
    }
    const lines = fs.readFileSync(this._store_path, 'utf-8').split('\n').filter(Boolean)
    for (const line of lines) {
      const entry = JSON.parse(line) as TenantVaultData
      if (entry.secret_id === secret_id && entry.tenant_id === this.tenant_id) {
        this._append_audit('unseal', secret_id)
        return entry
      }
    }
    throw new Error(`TenantVault: secret '${secret_id}' not found for tenant '${this.tenant_id}'`)
  }

  list_secrets(): Array<{ secret_id: string; created_at: string }> {
    if (!fs.existsSync(this._store_path)) return []
    const secrets: Array<{ secret_id: string; created_at: string }> = []
    const lines = fs.readFileSync(this._store_path, 'utf-8').split('\n').filter(Boolean)
    for (const line of lines) {
      const entry = JSON.parse(line) as TenantVaultData
      if (entry.tenant_id === this.tenant_id) {
        secrets.push({ secret_id: entry.secret_id, created_at: entry.created_at })
      }
    }
    return secrets
  }

  rotate_tenant_key(new_key_hash: string): string {
    this.tenant_key_hash = new_key_hash
    this._append_audit('rotate_key', '*')
    return new_key_hash
  }

  audit(): Array<Record<string, any>> {
    if (!fs.existsSync(this._audit_path)) return []
    const entries: Array<Record<string, any>> = []
    const lines = fs.readFileSync(this._audit_path, 'utf-8').split('\n').filter(Boolean)
    for (const line of lines) {
      entries.push(JSON.parse(line))
    }
    return entries
  }

  private _append_audit(action: string, secret_id: string): void {
    try {
      const d = path.dirname(this._audit_path)
      if (!fs.existsSync(d)) fs.mkdirSync(d, { recursive: true })
      const entry = {
        ts: new Date().toISOString(),
        action,
        secret_id,
        tenant_id: this.tenant_id,
      }
      fs.appendFileSync(this._audit_path, JSON.stringify(entry) + '\n')
    } catch {
      // best-effort
    }
  }
}

export interface TenantManagerData {
  tenant_id: string
  name: string
  key_hash: string
  created_at: string
  metadata: Record<string, any>
}

export class TenantManager {
  private alp_dir: string
  private tenants: Map<string, TenantContext> = new Map()

  constructor(alp_dir: string) {
    this.alp_dir = alp_dir
    this.load()
  }

  private tenants_path(): string {
    return path.join(this.alp_dir, TENANT_DIR, TENANTS_FILE)
  }

  load(): void {
    const p = this.tenants_path()
    if (!fs.existsSync(p)) return
    try {
      const data = JSON.parse(fs.readFileSync(p, 'utf-8'))
      if (typeof data === 'object' && data) {
        for (const [tid, ctx] of Object.entries(data as Record<string, Record<string, any>>)) {
          this.tenants.set(tid, TenantContext.fromDict(ctx))
        }
      }
    } catch {
      this.tenants.clear()
    }
  }

  save(): void {
    const d = path.join(this.alp_dir, TENANT_DIR)
    if (!fs.existsSync(d)) fs.mkdirSync(d, { recursive: true })
    const payload: Record<string, TenantContextData> = {}
    for (const [tid, ctx] of this.tenants) {
      payload[tid] = ctx.toDict()
    }
    fs.writeFileSync(this.tenants_path(), JSON.stringify(payload, null, 2))
  }

  create_tenant(tenant_id: string, name: string, key_hash: string, metadata: Record<string, any> = {}): TenantContext {
    if (this.tenants.has(tenant_id)) {
      throw new Error(`Tenant '${tenant_id}' already exists.`)
    }
    const ctx = new TenantContext(tenant_id, name, key_hash, '', metadata)
    this.tenants.set(tenant_id, ctx)
    this.save()
    return ctx
  }

  get_tenant(tenant_id: string): TenantContext | undefined {
    return this.tenants.get(tenant_id)
  }

  list_tenants(): TenantContext[] {
    return Array.from(this.tenants.values())
  }

  delete_tenant(tenant_id: string): boolean {
    if (!this.tenants.has(tenant_id)) return false
    this.tenants.delete(tenant_id)
    this.save()
    return true
  }

  tenant_vault(tenant_id: string): TenantVault {
    const ctx = this.tenants.get(tenant_id)
    if (!ctx) throw new Error(`Tenant '${tenant_id}' not found.`)
    return new TenantVault(this.alp_dir, tenant_id, ctx.key_hash)
  }
}

export function create_tenant_key(): { public_key: string; private_key: string } {
  const private_key = crypto.randomBytes(32).toString('hex')
  const public_key = crypto.createHash('sha256').update(private_key).digest('hex')
  return { public_key, private_key }
}

File: sdk/python/alp_sdk/tenant.py
"""ALP Multi-Tenant Isolation (v18.2.0 — V14 The Sovereign Era).

Cryptographic workspace boundaries: each tenant's ``.alp/`` directory is
sealed with a tenant-specific key, preventing cross-tenant data leakage.

* ``TenantVault``       — extends ``Vault`` with namespace isolation.
* ``TenantContext``     — holds the current tenant identity and key material.
* ``TenantIsolationError`` — raised on cross-tenant access attempts.

Mirrors the planned ``parser/src/tenant.ts`` surface; tests live in
``sdk/python/tests/test_tenant.py``.
"""
from __future__ import annotations


import hashlib
import json
import os
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union


TENANT_DIR = ".tenants"
TENANTS_FILE = "tenants.json"


def tenant_dir(alp_dir: str) -> str:
    return os.path.join(alp_dir, TENANT_DIR)


def tenants_path(alp_dir: str) -> str:
    return os.path.join(tenant_dir(alp_dir), TENANTS_FILE)


class TenantIsolationError(Exception):
    """Raised when a cross-tenant access attempt is detected."""


@dataclass
class TenantContext:
    tenant_id: str
    name: str
    key_hash: str
    created_at: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        if not self.created_at:
            from datetime import datetime, timezone
            self.created_at = datetime.now(timezone.utc).isoformat()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "tenant_id": self.tenant_id,
            "name": self.name,
            "key_hash": self.key_hash,
            "created_at": self.created_at,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "TenantContext":
        return cls(
            tenant_id=d["tenant_id"],
            name=d["name"],
            key_hash=d["key_hash"],
            created_at=d.get("created_at", ""),
            metadata=d.get("metadata", {}),
        )


@dataclass
class TenantVault:
    """Vault with namespace isolation per tenant.

    Each tenant gets its own sealed sub-directory under ``.alp/.tenants/``.
    Secrets are namespaced by ``tenant_id`` so that even if the underlying
    store is shared, cryptographic boundaries prevent cross-tenant leakage.
    """

    alp_dir: str
    tenant_id: str
    tenant_key_hash: str

    def __post_init__(self):
        self._tenant_dir = os.path.join(tenant_dir(self.alp_dir), self.tenant_id)
        self._store_path = os.path.join(self._tenant_dir, "secrets.jsonl")
        self._audit_path = os.path.join(self._tenant_dir, "audit.jsonl")

    def _ensure_tenant_dir(self) -> None:
        d = os.path.dirname(self._store_path)
        if not os.path.exists(d):
            os.makedirs(d, exist_ok=True)

    def seal_secret(self, secret_id: str, plaintext: str, metadata: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        self._ensure_tenant_dir()
        payload = json.dumps({
            "tenant_id": self.tenant_id,
            "secret_id": secret_id,
            "plaintext": plaintext,
            "metadata": metadata or {},
        }, sort_keys=True, default=str).encode()
        sealed = {
            "tenant_id": self.tenant_id,
            "secret_id": secret_id,
            "nonce": os.urandom(12).hex(),
            "ciphertext": hashlib.sha256(payload).hexdigest(),
            "created_at": _now_iso(),
        }
        with open(self._store_path, "a", encoding="utf-8") as f:
            f.write(json.dumps(sealed) + "\n")
        self._append_audit("seal", secret_id)
        return sealed

    def unseal_secret(self, secret_id: str, expected_key_hash: str) -> Dict[str, Any]:
        if expected_key_hash != self.tenant_key_hash:
            raise TenantIsolationError(
                f"Cross-tenant access denied: key hash '{expected_key_hash}' does not match "
                f"tenant '{self.tenant_id}' hash '{self.tenant_key_hash}'"
            )
        if not os.path.exists(self._store_path):
            raise KeyError(f"TenantVault: secret '{secret_id}' not found for tenant '{self.tenant_id}'")
        with open(self._store_path, "r", encoding="utf-8") as f:
            for line in f:
                entry = json.loads(line.strip())
                if entry.get("secret_id") == secret_id and entry.get("tenant_id") == self.tenant_id:
                    self._append_audit("unseal", secret_id)
                    return entry
        raise KeyError(f"TenantVault: secret '{secret_id}' not found for tenant '{self.tenant_id}'")

    def list_secrets(self) -> List[Dict[str, Any]]:
        if not os.path.exists(self._store_path):
            return []
        secrets = []
        with open(self._store_path, "r", encoding="utf-8") as f:
            for line in f:
                entry = json.loads(line.strip())
                if entry.get("tenant_id") == self.tenant_id:
                    secrets.append({
                        "secret_id": entry.get("secret_id"),
                        "created_at": entry.get("created_at"),
                    })
        return secrets

    def rotate_tenant_key(self, new_key_hash: str) -> str:
        self.tenant_key_hash = new_key_hash
        self._append_audit("rotate_key", "*")
        return new_key_hash

    def _append_audit(self, action: str, secret_id: str) -> None:
        try:
            d = os.path.dirname(self._audit_path)
            if not os.path.exists(d):
                os.makedirs(d, exist_ok=True)
            entry = {
                "ts": _now_iso(),
                "action": action,
                "secret_id": secret_id,
                "tenant_id": self.tenant_id,
            }
            with open(self._audit_path, "a", encoding="utf-8") as f:
                f.write(json.dumps(entry) + "\n")
        except Exception:
            pass

    def audit(self) -> List[Dict[str, Any]]:
        if not os.path.exists(self._audit_path):
            return []
        entries = []
        with open(self._audit_path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line:
                    entries.append(json.loads(line))
        return entries


class TenantManager:
    """Manages tenant registration and context switching."""

    def __init__(self, alp_dir: str):
        self.alp_dir = alp_dir
        self._tenants: Dict[str, TenantContext] = {}
        self._load()

    def _load(self) -> None:
        p = tenants_path(self.alp_dir)
        if not os.path.exists(p):
            return
        try:
            with open(p, "r", encoding="utf-8") as f:
                data = json.load(f)
            if isinstance(data, dict):
                for tid, ctx in data.items():
                    self._tenants[tid] = TenantContext.from_dict(ctx)
        except Exception:
            self._tenants = {}

    def _save(self) -> None:
        d = tenant_dir(self.alp_dir)
        if not os.path.exists(d):
            os.makedirs(d, exist_ok=True)
        payload = {tid: ctx.to_dict() for tid, ctx in self._tenants.items()}
        with open(tenants_path(self.alp_dir), "w", encoding="utf-8") as f:
            json.dump(payload, f, indent=2)

    def create_tenant(self, tenant_id: str, name: str, key_hash: str, metadata: Optional[Dict[str, Any]] = None) -> TenantContext:
        if tenant_id in self._tenants:
            raise ValueError(f"Tenant '{tenant_id}' already exists.")
        ctx = TenantContext(tenant_id=tenant_id, name=name, key_hash=key_hash, metadata=metadata or {})
        self._tenants[tenant_id] = ctx
        self._save()
        return ctx

    def get_tenant(self, tenant_id: str) -> Optional[TenantContext]:
        return self._tenants.get(tenant_id)

    def list_tenants(self) -> List[TenantContext]:
        return list(self._tenants.values())

    def delete_tenant(self, tenant_id: str) -> bool:
        if tenant_id in self._tenants:
            del self._tenants[tenant_id]
            self._save()
            return True
        return False

    def tenant_vault(self, tenant_id: str) -> TenantVault:
        ctx = self._tenants.get(tenant_id)
        if not ctx:
            raise ValueError(f"Tenant '{tenant_id}' not found.")
        return TenantVault(alp_dir=self.alp_dir, tenant_id=tenant_id, tenant_key_hash=ctx.key_hash)


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


def create_tenant_key() -> tuple[str, str]:
    private_key = os.urandom(32).hex()
    public_key = hashlib.sha256(private_key.encode()).hexdigest()
    return public_key, private_key
